### PR TITLE
[SPARK-48059][CORE] Implement the structured log framework on the java side

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to implement the structured log framework on the `java side`.

### Why are the changes needed?
Currently, the structured log framework on the `scala side` is basically available, but the`Spark Core` code also includes some `Java code`, which also needs to be connected to the structured log framework.

### Does this PR introduce _any_ user-facing change?
Yes, only for developers.

### How was this patch tested?
- Add some new UT.
- Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by panbingkun** [comment]
@gengliangwang 
In the migration of logs to the structured log framework, I found that we `missed` the corresponding implementation on the `java side`. This PR is to supplement it.

**Comment by panbingkun** [comment]
We currently support three ways to write logs:
- loggger.{error/warn/info/debug/trace}(string)
- loggger.{error/warn/info/debug/trace}(string, throwable)
-  loggger.{error/warn/info/debug/trace}({string template with {} inside}, JMDC...), eg:
  logger.info("Executor {} failed", JMDC(EXECUTOR_ID, "xxxx")

**Comment by gengliangwang** [comment]
@panbingkun Thanks for bringing this up.
I check the current Spark java code:
```
find . -name "*.java"|xargs grep "logInfo\|logWarn\|logError"|grep -v target|grep -v test
```
and found no log4j usages.  Is this for external systems? If yes, the external system should be able to use the current Logging trait as well, right?

**Comment by panbingkun** [comment]
> @panbingkun Thanks for bringing this up. I check the current Spark java code:
> 
> ```
> find . -name "*.java"|xargs grep "logInfo\|logWarn\|logError"|grep -v target|grep -v test
> ```
> 
> and found no log4j usages. Is this for external systems? If yes, the external system should be able to use the current Logging trait as well, right?

@gengliangwang 
Hmm..., let me give you some examples, eg:
https://github.com/apache/spark/blob/9caa6f7f8b8e9d906bc4a565179687350bfc7bb3/core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java#L285-L290

https://github.com/apache/spark/blob/9caa6f7f8b8e9d906bc4a565179687350bfc7bb3/core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java#L395

https://github.com/apache/spark/blob/9caa6f7f8b8e9d906bc4a565179687350bfc7bb3/connector/spark-ganglia-lgpl/src/main/java/com/codahale/metrics/ganglia/GangliaReporter.java#L336

image

**Comment by gengliangwang** [comment]
@panbingkun I see. There are about 88 loggings with variables
```
find . -name "*.java"|xargs grep -i "logger.info\|logger.warn\|logger.error"|grep "{}"| grep -v target|grep -v test
```
Since all the logs are using `{}` as variable placeholders, shall we try following the approach `Sequence of Key-Value Pairs (Seq[(K, V)])` from the [SPIP](https://docs.google.com/document/d/1rATVGmFLNVLmtxSpWrEceYm7d-ocgu8ofhryVs4g3XU/edit?usp=sharing)?
Also, I think we can reference the LogKey.scala in the java side.

**Comment by panbingkun** [comment]
> @panbingkun I see. There are about 88 loggings with variables
> 
> ```
> find . -name "*.java"|xargs grep -i "logger.info\|logger.warn\|logger.error"|grep "{}"| grep -v target|grep -v test
> ```
> 
> Since all the logs are using `{}` as variable placeholders, shall we try following the approach `Sequence of Key-Value Pairs (Seq[(K, V)])` from the [SPIP](https://docs.google.com/document/d/1rATVGmFLNVLmtxSpWrEceYm7d-ocgu8ofhryVs4g3XU/edit?usp=sharing)? Also, I think we can reference the LogKey.scala in the java side.

I think this is a good suggestion. Let me try it.

File: common/utils/src/main/java/org/apache/spark/internal/Logger.java [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    }
+  }
+
+  public void error(String msg, Throwable throwable) {
+    if (slf4jLogger.isErrorEnabled()) {
+      slf4jLogger.error(msg, throwable);
+    }
+  }
+
+  public void error(String msg, MDC... mdcs) {
```
Comment (panbingkun) [inline common/utils/src/main/java/org/apache/spark/internal/Logger.java:49]: Although the following design is shown in the document, 
image
But I think the parameter `void error(String msg, MDC... mdcs)` is more in line with the style of Java? WDYT @gengliangwang

Comment (gengliangwang) [reply common/utils/src/main/java/org/apache/spark/internal/Logger.java:49]: Either way is fine to me

**Comment by panbingkun** [comment]
@gengliangwang
This pr is ready for review.

File: common/utils/src/main/java/org/apache/spark/internal/Logger.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+  public void warn(String msg, Throwable throwable) {
+    slf4jLogger.warn(msg, throwable);
+  }
+
+  public void warn(String msg, MDC... mdcs) {
+    if (mdcs == null || mdcs.length == 0) {
+      slf4jLogger.warn(msg);
+    } else {
+      withLogContext(msg, mdcs, null, mt -> slf4jLogger.warn(mt.message));
```
Comment (gengliangwang) [inline common/utils/src/main/java/org/apache/spark/internal/Logger.java:73]: Let's check the log level before calling `withLogContext`. For example, if the log level is ERROR, we don't need to enter the method `withLogContext`

Comment (panbingkun) [reply common/utils/src/main/java/org/apache/spark/internal/Logger.java:73]: Done

**Review by gengliangwang — APPROVED** [review]

File: common/utils/src/test/java/org/apache/spark/util/LoggerSuiteBase.java [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.logging.log4j.Level;
+import org.junit.jupiter.api.Test;
+
+import org.apache.spark.internal.Logger;
+import org.apache.spark.internal.LogKeys;
+import org.apache.spark.internal.MDC;
+
+public abstract class LoggerSuiteBase {
```
Comment (gengliangwang) [inline common/utils/src/test/java/org/apache/spark/util/LoggerSuiteBase.java:33]: nit: ideally, we can have a followup to reduce the duplicated test code.

**Comment by gengliangwang** [comment]
Thanks, merging to master

**Comment by panbingkun** [comment]
> Thanks, merging to master

Thank you very much.
Based on this, I will do log migration and auxiliary work on the java side.
(eg, after the migration is completed, we can add a `rule` in `checkstyle` that prohibits importing `org.slf4j.Logger` and `org.slf4j.LoggerFactory`)

**Comment by gengliangwang** [comment]
@panbingkun Awesome, thanks for the work!

File: common/utils/src/main/java/org/apache/spark/internal/Logger.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    if (mdcs == null || mdcs.length == 0) {
+      slf4jLogger.warn(msg);
+    } else if (slf4jLogger.isWarnEnabled()) {
+      withLogContext(msg, mdcs, null, mt -> slf4jLogger.warn(mt.message));
+    }
+  }
+
+  public void warn(String msg, Throwable throwable, MDC... mdcs) {
+    if (mdcs == null || mdcs.length == 0) {
+      slf4jLogger.warn(msg);
```
Comment (gengliangwang) [inline common/utils/src/main/java/org/apache/spark/internal/Logger.java:79]: We should have the `throwable` here. Also, we need test case for this.

File: common/utils/src/main/java/org/apache/spark/internal/Logger.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    if (mdcs == null || mdcs.length == 0) {
+      slf4jLogger.info(msg);
+    } else if (slf4jLogger.isInfoEnabled()) {
+      withLogContext(msg, mdcs, null, mt -> slf4jLogger.info(mt.message));
+    }
+  }
+
+  public void info(String msg, Throwable throwable, MDC... mdcs) {
+    if (mdcs == null || mdcs.length == 0) {
+      slf4jLogger.info(msg);
```
Comment (gengliangwang) [inline common/utils/src/main/java/org/apache/spark/internal/Logger.java:103]: We should have the `throwable` here. Also, we need test case for this.

File: common/utils/src/main/java/org/apache/spark/internal/Logger.java (hunk 4 above)
Comment (panbingkun) [reply common/utils/src/main/java/org/apache/spark/internal/Logger.java:79]: Okay, let me do it.

File: common/utils/src/main/java/org/apache/spark/internal/Logger.java (hunk 5 above)
Comment (panbingkun) [reply common/utils/src/main/java/org/apache/spark/internal/Logger.java:103]: Okay

Comment (panbingkun) [reply common/utils/src/main/java/org/apache/spark/internal/Logger.java:103]: @gengliangwang 
I have submitted a PR to fix it: https://github.com/apache/spark/pull/47095

**Pull request closed without merging**